### ISSUE TITLE
DSS fails inference

### ISSUE THREAD

**Issue by charleskawczynski** [CliMA/ClimaCore.jl#1283]
## Describe the bug
dss fails inference:
```julia
julia> JET.@test_opt CA.dss!(Yₜ, p, t)
JET-test failed at REPL[66]:1
  Expression: #= REPL[66]:1 =# JET.@test_opt CA.dss!(Yₜ, p, t)
  ═════ 4 possible errors found ═════
  ┌ @ /Users/charliekawczynski/Dropbox/Caltech/work/dev/CliMA/ClimaAtmos.jl/src/prognostic_equations/dss.jl:17 ClimaCore.Spaces.weighted_dss_start2!(Y.c, p.ghost_buffer.c)
  │┌ @ /Users/charliekawczynski/.julia/packages/ClimaCore/A0lxf/src/Fields/Fields.jl:400 ClimaCore.Spaces.weighted_dss_start!(field, ghost_buffer)
  ││┌ @ /Users/charliekawczynski/.julia/packages/ClimaCore/A0lxf/src/Fields/Fields.jl:366 ClimaCore.Spaces.weighted_dss_start!(ClimaCore.Fields.field_values(field), ClimaCore.Fields.axes(field), dss_buffer)
  │││┌ @ /Users/charliekawczynski/.julia/packages/ClimaCore/A0lxf/src/Spaces/dss.jl:261 ClimaCore.Spaces.weighted_dss_start!(data, space, ClimaCore.Spaces.horizontal_space(space), dss_buffer)
  ││││┌ @ /Users/charliekawczynski/.julia/packages/ClimaCore/A0lxf/src/Spaces/dss.jl:282 ClimaCore.Spaces.dss_local_ghost!(device, dss_buffer.perimeter_data, ClimaCore.Spaces.perimeter(hspace), hspace.topology)
  │││││┌ @ /Users/charliekawczynski/.julia/packages/ClimaCore/A0lxf/src/Spaces/dss.jl:874 zero_data = ClimaCore.Spaces.map(ClimaCore.Spaces.zero, ClimaCore.Spaces.slab(perimeter_data, 1, 1)[1])
  ││││││┌ @ namedtuple.jl:219 map(tuple(f), map(Base.Tuple, tuple(tuple(nt), nts...))...)
  │││││││┌ @ tuple.jl:224 map(f, Base.tail(t))
  ││││││││┌ @ tuple.jl:224 map(f, Base.tail(t))
  │││││││││┌ @ tuple.jl:224 map(f, Base.tail(t))
  ││││││││││┌ @ tuple.jl:223 f(%1)
  │││││││││││ runtime dispatch detected: f::typeof(zero)(%1::NamedTuple{(:ρatke,), Tuple{Float32}})
  ││││││││││└────────────────
  ┌ @ /Users/charliekawczynski/Dropbox/Caltech/work/dev/CliMA/ClimaAtmos.jl/src/prognostic_equations/dss.jl:18 ClimaCore.Spaces.weighted_dss_start2!(Y.f, p.ghost_buffer.f)
  │┌ @ /Users/charliekawczynski/.julia/packages/ClimaCore/A0lxf/src/Fields/Fields.jl:400 ClimaCore.Spaces.weighted_dss_start!(field, ghost_buffer)
  ││┌ @ /Users/charliekawczynski/.julia/packages/ClimaCore/A0lxf/src/Fields/Fields.jl:366 ClimaCore.Spaces.weighted_dss_start!(ClimaCore.Fields.field_values(field), ClimaCore.Fields.axes(field), dss_buffer)
  │││┌ @ /Users/charliekawczynski/.julia/packages/ClimaCore/A0lxf/src/Spaces/dss.jl:261 ClimaCore.Spaces.weighted_dss_start!(data, space, ClimaCore.Spaces.horizontal_space(space), dss_buffer)
  ││││┌ @ /Users/charliekawczynski/.julia/packages/ClimaCore/A0lxf/src/Spaces/dss.jl:282 ClimaCore.Spaces.dss_local_ghost!(device, dss_buffer.perimeter_data, ClimaCore.Spaces.perimeter(hspace), hspace.topology)
  │││││┌ @ /Users/charliekawczynski/.julia/packages/ClimaCore/A0lxf/src/Spaces/dss.jl:874 zero_data = ClimaCore.Spaces.map(ClimaCore.Spaces.zero, ClimaCore.Spaces.slab(perimeter_data, 1, 1)[1])
  ││││││┌ @ namedtuple.jl:219 map(tuple(f), map(Base.Tuple, tuple(tuple(nt), nts...))...)
  │││││││┌ @ tuple.jl:222 f(%1)
  ││││││││ runtime dispatch detected: f::typeof(zero)(%1::Tuple{NamedTuple{(:u₃,), Tuple{ClimaCore.Geometry.Covariant3Vector{Float32}}}})
  │││││││└────────────────
  ┌ @ /Users/charliekawczynski/Dropbox/Caltech/work/dev/CliMA/ClimaAtmos.jl/src/prognostic_equations/dss.jl:19 ClimaCore.Spaces.weighted_dss_internal2!(Y.c, p.ghost_buffer.c)
  │┌ @ /Users/charliekawczynski/.julia/packages/ClimaCore/A0lxf/src/Fields/Fields.jl:403 ClimaCore.Spaces.weighted_dss_internal!(field, ghost_buffer)
  ││┌ @ /Users/charliekawczynski/.julia/packages/ClimaCore/A0lxf/src/Fields/Fields.jl:369 ClimaCore.Spaces.weighted_dss_internal!(ClimaCore.Fields.field_values(field), ClimaCore.Fields.axes(field), dss_buffer)
  │││┌ @ /Users/charliekawczynski/.julia/packages/ClimaCore/A0lxf/src/Spaces/dss.jl:325 ClimaCore.Spaces.weighted_dss_internal!(data, space, ClimaCore.Spaces.horizontal_space(space), dss_buffer)
  ││││┌ @ /Users/charliekawczynski/.julia/packages/ClimaCore/A0lxf/src/Spaces/dss.jl:356 ClimaCore.Spaces.dss_local!(device, dss_buffer.perimeter_data, ClimaCore.Spaces.perimeter(hspace), hspace.topology)
  │││││┌ @ /Users/charliekawczynski/.julia/packages/ClimaCore/A0lxf/src/Spaces/dss.jl:796 ClimaCore.Spaces.dss_local_vertices!(perimeter_data, perimeter, topology)
  ││││││┌ @ /Users/charliekawczynski/.julia/packages/ClimaCore/A0lxf/src/Spaces/dss.jl:811 sum_data = ClimaCore.Spaces.mapreduce(#16, ClimaCore.Spaces.:⊞, vertex)
  │││││││┌ @ reduce.jl:302 Base.:(var"#mapreduce#263")(pairs(NamedTuple()), #self#, f, op, itr)
  ││││││││┌ @ reduce.jl:302 mapfoldl(f, op, itr)
  │││││││││┌ @ reduce.jl:170 Base.:(var"#mapfoldl#259")(Base._InitialValue(), #self#, f, op, itr)
  ││││││││││┌ @ reduce.jl:170 Base.mapfoldl_impl(f, op, init, itr)
  │││││││││││┌ @ reduce.jl:44 Base.foldl_impl(op′, nt, itr′)
  ││││││││││││┌ @ reduce.jl:49 Base.reduce_empty_iter(op, itr)
  │││││││││││││┌ @ reduce.jl:378 Base.reduce_empty_iter(op, itr, Base.IteratorEltype(itr))
  ││││││││││││││┌ @ reduce.jl:379 Base.reduce_empty(op, eltype(itr))
  │││││││││││││││┌ @ reduce.jl:356 Base.mapreduce_empty(%1, [quote], T)
  ││││││││││││││││ runtime dispatch detected: Base.mapreduce_empty(%1::ClimaCore.Spaces.var"#16#17"{ClimaCore.DataLayouts.VIFH{NamedTuple{(:ρ, :uₕ, :ρe_tot, :ρq_tot, :sgs⁰, :sgsʲs), Tuple{Float32, ClimaCore.Geometry.UVVector{Float32}, Float32, Float32, NamedTuple{(:ρatke,), Tuple{Float32}}, Tuple{NamedTuple{(:ρa, :ρae_tot, :ρaq_tot), Tuple{Float32, Float32, Float32}}}}}, 12, Array{Float32, 4}}, Int64}, [quote]::Base.BottomRF{typeof(ClimaCore.RecursiveApply.radd)}, T::Type{Tuple{Int64, Int64}})
  │││││││││││││││└─────────────────
  ┌ @ /Users/charliekawczynski/Dropbox/Caltech/work/dev/CliMA/ClimaAtmos.jl/src/prognostic_equations/dss.jl:20 ClimaCore.Spaces.weighted_dss_internal2!(Y.f, p.ghost_buffer.f)
  │┌ @ /Users/charliekawczynski/.julia/packages/ClimaCore/A0lxf/src/Fields/Fields.jl:403 ClimaCore.Spaces.weighted_dss_internal!(field, ghost_buffer)
  ││┌ @ /Users/charliekawczynski/.julia/packages/ClimaCore/A0lxf/src/Fields/Fields.jl:369 ClimaCore.Spaces.weighted_dss_internal!(ClimaCore.Fields.field_values(field), ClimaCore.Fields.axes(field), dss_buffer)
  │││┌ @ /Users/charliekawczynski/.julia/packages/ClimaCore/A0lxf/src/Spaces/dss.jl:325 ClimaCore.Spaces.weighted_dss_internal!(data, space, ClimaCore.Spaces.horizontal_space(space), dss_buffer)
  ││││┌ @ /Users/charliekawczynski/.julia/packages/ClimaCore/A0lxf/src/Spaces/dss.jl:356 ClimaCore.Spaces.dss_local!(device, dss_buffer.perimeter_data, ClimaCore.Spaces.perimeter(hspace), hspace.topology)
  │││││┌ @ /Users/charliekawczynski/.julia/packages/ClimaCore/A0lxf/src/Spaces/dss.jl:796 ClimaCore.Spaces.dss_local_vertices!(perimeter_data, perimeter, topology)
  ││││││┌ @ /Users/charliekawczynski/.julia/packages/ClimaCore/A0lxf/src/Spaces/dss.jl:811 sum_data = ClimaCore.Spaces.mapreduce(#16, ClimaCore.Spaces.:⊞, vertex)
  │││││││┌ @ reduce.jl:302 Base.:(var"#mapreduce#263")(pairs(NamedTuple()), #self#, f, op, itr)
  ││││││││┌ @ reduce.jl:302 mapfoldl(f, op, itr)
  │││││││││┌ @ reduce.jl:170 Base.:(var"#mapfoldl#259")(Base._InitialValue(), #self#, f, op, itr)
  ││││││││││┌ @ reduce.jl:170 Base.mapfoldl_impl(f, op, init, itr)
  │││││││││││┌ @ reduce.jl:44 Base.foldl_impl(op′, nt, itr′)
  ││││││││││││┌ @ reduce.jl:49 Base.reduce_empty_iter(op, itr)
  │││││││││││││┌ @ reduce.jl:378 Base.reduce_empty_iter(op, itr, Base.IteratorEltype(itr))
  ││││││││││││││┌ @ reduce.jl:379 Base.reduce_empty(op, eltype(itr))
  │││││││││││││││┌ @ reduce.jl:356 Base.mapreduce_empty(%1, [quote], T)
  ││││││││││││││││ runtime dispatch detected: Base.mapreduce_empty(%1::ClimaCore.Spaces.var"#16#17"{ClimaCore.DataLayouts.VIFH{NamedTuple{(:u₃, :sgsʲs), Tuple{ClimaCore.Geometry.Covariant3Vector{Float32}, Tuple{NamedTuple{(:u₃,), Tuple{ClimaCore.Geometry.Covariant3Vector{Float32}}}}}}, 12, Array{Float32, 4}}, Int64}, [quote]::Base.BottomRF{typeof(ClimaCore.RecursiveApply.radd)}, T::Type{Tuple{Int64, Int64}})
  │││││││││││││││└─────────────────
  
ERROR: There was an error during testing
```

## Related issues / PRs

Relavent PR: https://github.com/CliMA/ClimaCore.jl/pull/1198

**Comment by charleskawczynski** [comment]
Closed by #1198.

**Comment by charleskawczynski** [comment]
New failure in https://buildkite.com/clima/climaatmos-ci/builds/10182#01888df6-f0aa-4027-aa12-2e5f860da748

**Comment by charleskawczynski** [comment]
Another example found in #1311.

**Comment by charleskawczynski** [comment]
Closed by #1334.
